fix: add support for deserialize a class extending ArrayList

suraj-qlogic opened the pull request.

suraj-qlogic committed:
fix: add support for deserialize a class extending ArrayList

athakor reviewed (commented):
Linkage monitor failure needs to be addressed.
```
SEVERE: Newly introduced problem:
(com.google.api:api-common:1.9.3) com.google.api.core.ApiFutures's method successfulAsList(Iterable) is not found
  referenced from com.google.cloud.firestore.BulkWriter (com.google.cloud:google-cloud-firestore:1.35.3-SNAPSHOT)

com.google.api:api-common:1.9.3 is at:
  com.google.api:api-common:1.9.3 (compile)
  and 47 other dependency paths.
com.google.cloud:google-cloud-firestore:1.35.3-SNAPSHOT is at:
  com.google.cloud:google-cloud-firestore:1.35.3-SNAPSHOT (compile)
```

BenWhitehead committed:
chore(build): add maven profile to enable samples (#329)

Commit:
chore: make BulkWriter package-private (#330)

Co-authored-by: BenWhitehead <BenWhitehead@users.noreply.github.com>

release-please[bot] committed:
chore: release 2.0.0 (#309)

:robot: I have created a release \*beep\* \*boop\* 
---
## [2.0.0](https://www.github.com/googleapis/java-firestore/compare/v1.35.2...v2.0.0) (2020-08-14)


### New Features

#### Query Partition API

New API and backend RPC which allows for fetching a set of cursor keys for a 
Collection Group Query. Accessible via the new [`CollectionGroup#getPartitions(long,ApiStreamObserver)`](https://googleapis.dev/java/google-cloud-firestore/2.0.0/com/google/cloud/firestore/CollectionGroup.html#getPartitions-long-com.google.api.gax.rpc.ApiStreamObserver-) method. 

#### Read-Only Transaction Options

[`TransactionOptions`](https://googleapis.dev/java/google-cloud-firestore/2.0.0/com/google/cloud/firestore/TransactionOptions.html)
has been refactored to provide the ability to configure options for read-only 
transactions along with the existing configuration for read-write transactions.

This new ability is provided via the new [`TransactionOptions.createReadOnlyOptionsBuilder()`](https://googleapis.dev/java/google-cloud-firestore/2.0.0/com/google/cloud/firestore/TransactionOptions.html#createReadOnlyOptionsBuilder--) 
type safe builder.

Along with the new type safe builder for read-only options, there is a new type 
safe builder for read-write options as well accessible via [`TransactionOptions.createReadWriteOptionsBuilder()`](https://googleapis.dev/java/google-cloud-firestore/2.0.0/com/google/cloud/firestore/TransactionOptions.html#createReadWriteOptionsBuilder--). Each of the existing `TransactionOptions.create(...)`
methods for configuring read-write options has been deprecated in favor of the new builder.

#### EmulatorCredentials

`com.google.cloud.firestore.FirestoreOptions.Builder.FakeCredentials` has been
made static and renamed to `com.google.cloud.firestore.FirestoreOptions.EmulatorCredentials`
allowing instantiation outside `FirestoreOptions.Builder`.

When connecting to the Cloud Firestore Emulator via `FirestoreOptions` rather than
the environment variable `FIRESTORE_EMULATOR_HOST`, a custom credential implementation
must be specified to allow various admin operations in the emulator. Previously
this required users to create their own implementation due to it not being 
possible to construct a `FakeCredential`. As part of this change, `EmulatorCredentials`
is static and therefore able to be constructed from any location.

### Breaking Changes

#### New Firestore Admin Client API Artifact

The Cloud Firestore Admin Client has been migrated to its own maven artifact `com.google.cloud:google-cloud-firestore-admin`
rather than being bundled in `com.google.cloud:google-cloud-firestore`. All 
packages and classes have retained their existing names.

The new artifact is included in the `com.google.cloud:google-cloud-firestore-bom`, 
`com.google.cloud:google-cloud-bom` and `com.google.cloud:libraries-bom` 
artifacts and is accessible by adding the new dependency to your `pom.xml` file:

```xml
<dependency>
  <groupId>com.google.cloud</groupId>
  <artifactId>google-cloud-firestore-admin</artifactId>
</dependency>
```

#### Removal of v1beta1

Cloud Firestore has been GA for some time now, and the `google-cloud-firestore` 
code base has been using the protos and generated classes for the v1 api since 
that time. As such, we will no longer be publishing artifacts for the deprecated
v1beta1 protos. All functionality from v1beta1 is present in v1, and all users
should update any code to use v1.

#### Removal of support for `java.util.Date` in Snapshots

It is no longer possible to configure the ability for `java.util.Date` to be
returned from `DocumentSnapshot.get(FieldPath)` or `DocumentSnapshot.getData()`
for properties which are stored as Timestamps in Cloud Firestore.

The default behavior has been to return `com.google.cloud.Timestamp` by default
for some time, and is now the only option. Any code that is dependent on the old
behavior must be updated to use Timestamps instead of Date.

### Laundry List of Pull Requests

#### ⚠ BREAKING CHANGES

* add support for the Query Partition API (#202)
  * `Firestore#collectionGroup(...)` has a new return type `CollectionGroup` 
    which requires any code that previously used the method be re-compiled to
    pick up the new signature. `CollectionGroup` extends `Query` and as such 
    does not require your code to be updated, only the compiled class files.
* move FirestoreAdminClient and associated classes to new artifact google-cloud-firestore-admin (#311)
* remove deprecated v1beta1 protos and grpc client (#305)
* remove deprecated FirestoreOptions#setTimestampsInSnapshotsEnabled (#308)
* remove deprecated getCollections() methods (#307)
* various renames due to generator changes

#### Features

* add support for read-only transactions in TransactionOptions ([#320](https://www.github.com/googleapis/java-firestore/issues/320)) ([c25dca3](https://www.github.com/googleapis/java-firestore/commit/c25dca3ed6ca0c156ec60569ebc9f3a481bd4fee))
* add support for the Query Partition API ([#202](https://www.github.com/googleapis/java-firestore/issues/202)) ([3996548](https://www.github.com/googleapis/java-firestore/commit/39965489cbc836af573e500d57007c88241d7eb6))


#### Bug Fixes

* refactor FakeCredentials ([#325](https://www.github.com/googleapis/java-firestore/issues/325)) ([269e62c](https://www.github.com/googleapis/java-firestore/commit/269e62c6b8031d48e7f2e282b09b5ffcfadae547)), closes [#190](https://www.github.com/googleapis/java-firestore/issues/190)


#### Dependencies

* update dependency com.google.cloud:google-cloud-shared-dependencies to v0.8.5 ([#322](https://www.github.com/googleapis/java-firestore/issues/322)) ([1b21350](https://www.github.com/googleapis/java-firestore/commit/1b21350c0bc4a21cee2b281f944cbd061b1f8898))
* update dependency com.google.cloud:google-cloud-shared-dependencies to v0.8.6 ([#324](https://www.github.com/googleapis/java-firestore/issues/324)) ([b945fdb](https://www.github.com/googleapis/java-firestore/commit/b945fdb04da76a1e007d012c809449c5a43bb990))
* update jackson dependencies to v2.11.2 ([#314](https://www.github.com/googleapis/java-firestore/issues/314)) ([15d68cd](https://www.github.com/googleapis/java-firestore/commit/15d68cd93ac1fd206895fd37155a9ba82b9196ca))


#### Miscellaneous Chores

* enable gapicv2 ([#188](https://www.github.com/googleapis/java-firestore/issues/188)) ([92224bc](https://www.github.com/googleapis/java-firestore/commit/92224bcd52aa88cc6eb1da28747de0535d776a0f))
* move FirestoreAdminClient and associated classes to new artifact google-cloud-firestore-admin ([#311](https://www.github.com/googleapis/java-firestore/issues/311)) ([03ef755](https://www.github.com/googleapis/java-firestore/commit/03ef755dd164e6f1ec749f3f985b913b5ae23d14))
* remove deprecated FirestoreOptions#setTimestampsInSnapshotsEnabled ([#308](https://www.github.com/googleapis/java-firestore/issues/308)) ([7255a42](https://www.github.com/googleapis/java-firestore/commit/7255a42bcee3a6938dd5fafaef3465f948f39600))
* remove deprecated getCollections() methods ([#307](https://www.github.com/googleapis/java-firestore/issues/307)) ([bb4ddf1](https://www.github.com/googleapis/java-firestore/commit/bb4ddf1ce3cc3bd2e06a4ad5097bd18060e4467b))
* remove deprecated v1beta1 protos and grpc client ([#305](https://www.github.com/googleapis/java-firestore/issues/305)) ([96adacb](https://www.github.com/googleapis/java-firestore/commit/96adacbf52ace27e54b7a210d7c73b46922fbcbd))
* add BulkWriter ([#323](https://www.github.com/googleapis/java-firestore/issues/323)) ([e7054df](https://www.github.com/googleapis/java-firestore/commit/e7054df79b4139fdfd0cc6aa0620fbfa1a10a6b0))
* make BulkWriter package private ([#330](https://github.com/googleapis/java-firestore/pull/330)) ([ef0869a](https://github.com/googleapis/java-firestore/commit/ef0869a7fa619bc15fef27ad90d41cb718cb981d))

---


This PR was generated with [Release Please](https://github.com/googleapis/release-please).

renovate-bot committed:
chore(deps): update dependency com.google.cloud:libraries-bom to v9 (#331)

release-please[bot] committed:
chore: release 2.0.1-SNAPSHOT (#332)

:robot: I have created a release \*beep\* \*boop\* 
---
### Updating meta-information for bleeding-edge SNAPSHOT release.
---


This PR was generated with [Release Please](https://github.com/googleapis/release-please).
